frontend/catalyst/api_extensions/pipeline.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains public API functions that provide control for the 
+user to input what MLIR compiler passes to run. 
+
+Currently, each pass has its own user-facing decorator. In the future, 
+a unified user interface for all the passes is necessary. 
+
+Note that the pass pipeline table does not need to modify the qnode in 
+any way. Its only purpose it to mark down the passes the user wants to 
+run on each qnode, and let send this information to 
+frontend/catalyst/compiler.py to handle the actual running of the passes. 
+"""
+
+import pennylane as qml
+
+
+## PASS PIPELINE TABLE ##
+class ACTIVE_PASSES:
+    """
+    A class that records the active compilation passes on a qnode.
+    Note that the ordering of passes of course matters.
+    """
+
+    def __init__(self):
+        self.table = {}
+
+    def __repr__(self):
+        return str(self.table)
+
+    def getTable(self):
+        """
+        Return the table object.
+        """
+        return self.table
+
+    def add_pass_on_qnode(self, qnode, pass_):
+        """
+        qnode (QNODE): the qnode object to compile
+        pass_ (str):   the compiler pass to run on this object.
+                       At each call of add_pass, the pass is added to the end
+                       of the current pipeline for that qnode.
+
+        """
+        if qnode not in self.table:
+            self.table[qnode] = [pass_]
+        else:
+            self.table[qnode].append(pass_)
+
+    def query(self, qnode):
+        """
+        Look up a qnode from the table and return the list of passes to run on it.
+        """
+        return self.table[qnode]
+
+
+active_passes = ACTIVE_PASSES()
```

Comment:
Could it be a property of the QJIT class or QNode as it works within the scope of qnodes? I wonder what the best way is to remove this globally defined active_passes :thinking:

Reply:
The use of this as a global also makes me uncomfortable. However this property shouldn't belong to QJIT or QNode (talked to Josh on this). See the UI epic linked in a comment above, but very quickly:

- This is a catalyst only feature (core PL doesn't have the idea of "mlir compiler passes"), so making it into a QNode option does not make sense. 
- If the option is in QJIT, then we lose the ability to control what passes to run on what cricuits, since we recommend calling `qjit` once on the entire workflow. e.g.
```python
@qjit(run_passes=True)
def workflow(..):
   @qnode(..)
   def circuit1(..):
      ..

   @qnode(..)
   def circuit2(..):
      ..
```
will run the pass on both qnodes, and this UI does not offer any control on, say, a user only wanting to run a pass on some subset of circuits.

Reply:
Implementation-wise I should think of something to get rid of this as a global though, that part's absolutely right.

Reply:
Thanks @paul0403 for the elaboration! Maybe here the use of global is not that bad then, giving it more thoughts, we could say global variables in a Python module are not really "global". Their visibility is limited to the module and this makes them a way less harmful :) 

The other way around is the consideration of adding a new abstraction (`class pipeline`) to wrap this variable and helper methods. But in Python, this seems to be equivalent with the scope of "module"s.

If you decide to keep this global variable, I will suggest declaring `active_passes` with `global` to make them easier to track down in the future.

Reply:
I quite struggle to find a way to avoid this global: all the API functions cannot be class methods (since they need to be top-level API calls), but they all need to modify a common table. Any attempts to circumvent this will just create another global whose a more abstraction layer up. e.g. 
```python
class pipeline:
   active_passes = ACTIVE_PASSES()
   def cancel_inverses_impl(...):
      active_passes.add_pass_on_qnode(...)
      ...
   def merge_rotations_impl(...):
      active_passes.add_pass_on_qnode(...)

pipeline = pipeline()
def cancel_inverses(...):
   pipeline.cancel_inverses_impl(..)

def merge_rotations(..):
   pipeline.merge_rotations_impl(...)
```
but this just means instead of having a `active_passes` global, we have a `pipeline` global. I would really like this global to be protected by something, but I guess maybe we should just mark it `global` and thank the python module visibility's own protection...